pkg/cnab/directory_parameter.go

```diff
@@ -0,0 +1,104 @@
+package cnab
+
+import (
+	"encoding/json"
+
+	"github.com/cnabio/cnab-go/bundle/definition"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/pkg/errors"
+)
+
+const (
+	DirectoryExtensionShortHand    = "directory-parameter"
+	DirectoryParameterExtensionKey = PorterExtensionsPrefix + DirectoryExtensionShortHand
+)
+
+// DirectoryParameterDefinition represents those parameter options
+// That apply exclusively to the directory parameter type
+type DirectoryParameterDefinition struct {
+	Writeable bool `yaml:"writeable,omitempty"`
+	// UID and GID should be ints, however 0 is the default value for int type
+	// But is also a realistic value for UID/GID thus we need to make the type interface
+	// To detect the case that the values weren't set
+	GID interface{} `yaml:"gid,omitempty" json:"gid,omitempty"`
+	UID interface{} `yaml:"uid,omitempty" json:"uid,omitempty"`
+}
+
+// MountParameterSource represents a parameter using a docker mount
+// As a its source with the provided options
+type MountParameterSourceDefn struct {
+	mount.Mount `yaml:",inline"`
+	Name        string `json:"name,omitempty" yaml:"name,omitempty"`
+}
+
+// DirectorySources represents the sources available to the directory parameter type
+// Currently only mount has been specified, but this could change in the future
+type DirectorySources struct {
+	Mount MountParameterSourceDefn `yaml:"mount,omitempty" json:"mount,omitempty"`
+}
+type DirectoryDetails struct {
```

Comment:
I haven't tried this but here's what I think will fit better with how the CNAB spec and Porter works:

* DirectoryDetails should have Name defined as well, or use a map[string]DirectoryDetails when storing the extension data and then key off of parameter name.
* DirectoryDefaults should not have DirectorySources or Kind.
* Do not create a new parameter source of type mount, I think we can get it to work without it and sources should be used for things that the bundle author knows about up-front (like using the output of a bundle as the parameter value), and bundle authors shouldn't be involved with whether or not a directory was injected using a docker mount, or through some other trick (like unpacking a gzip file into the specified directory).